### ISSUE TITLE
dosdebug: incorrectly disassembles "push word [ebx]" in 16-bit cs as "push dword [ebx]"

### ISSUE THREAD

**Issue by ecm-pushbx** [dosemu2/dosemu2#1165]
**Describe the bug**

When tracing in dosdebug through a 16-bit code segment, a push memory instruction with an a32 prefix (32-bit effective addressing) but without an o32 prefix is incorrectly disassembled as pushing a dword. Execution is not affected, it does properly push a word as expected.

**To Reproduce**

lDebug input:

```
ldebug/bin$ hg id
382e3c004b23 tip
ldebug/bin$ dosemu -dumb -K "$PWD" -E "ldebugu.com"
dosemu2 2.0pre8-20200519-1858-g9cd29a3c9
Configured: 2020-05-19 22:56:58 +0200
Please test against a recent version before reporting bugs and problems.
Get the latest code at http://stsp.github.io/dosemu2
Submit Bugs via https://github.com/stsp/dosemu2/issues
Ask for help in mail list: linux-msdos@vger.kernel.org

FreeDOS kernel - SVN (build 2042 OEM:0xfd) [compiled Sep 22 2017]
Kernel compatibility 7.10 - BORLANDC - FAT32 support

(C) Copyright 1995-2012 Pasquale J. Villani and The FreeDOS Project.
All Rights Reserved. This is free software and comes with ABSOLUTELY NO
WARRANTY; you can redistribute it and/or modify it under the terms of the
GNU General Public License as published by the Free Software Foundation;
either version 2, or (at your option) any later version.
C: HD1, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
D: HD2, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
E: HD3, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
F: HD4, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
EMUFS host file and print access available                                     
dosemu XMS 3.0 & UMB support enabled
dosemu EMS driver rev 0.8 installed.
dosemu CDROM driver installed (V0.2)
Kernel: allocated 30 Diskbuffers = 15960 Bytes in HMA
emufs: redirector enabled
BLASTER=A220 I5 D1 H5 P330 T6
MIDI=SYNTH:2 MAP:E MODE:0
Welcome to dosemu2!
    Build 2.0pre8-20200519-1858-g9cd29a3c9
About to Execute : ldebugu.com
About to Execute : ldebugu.com
-a
1939:0100 mov ebx, 200
1939:0106 mov dword [ebx], 44332211
1939:010E push word [bx]
1939:0110 push word [ebx]
1939:0113 push dword [bx]
1939:0116 push dword [ebx]
1939:011A nop
1939:011B int3
1939:011C nop
1939:011D 
-; set breakpoint in dosdebug
-g
ldebug/bin$ 
```

dosdebug input:

```
.dosemu/run$ dosdebug

DOSEMU Debugger V0.6 connected
- type ? to get help on commands -
system state: running
AX=0070  BX=98fe  CX=0001  DX=0540  SI=0a3a  DI=0001  SP=09fe  BP=0a2c
DS=00e9  ES=047b  FS=0000  GS=0000  FL=000b0ad6
CS:IP=ff10:01e8       SS:SP=00e9:09fe

ff10:01e8 F4               hlt

dosdebug> stop
system state: stopped
AX=0070  BX=98fe  CX=0001  DX=0540  SI=0a3a  DI=0001  SP=09fe  BP=0a2c
DS=00e9  ES=047b  FS=0000  GS=0000  FL=00030ad6
CS:IP=ff10:01e8       SS:SP=00e9:09fe

ff10:01e8 F4               hlt

dosdebug> bp 1939:100
dosdebug> g
system state: stopped
AX=0000  BX=0000  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffe  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0246
CS:IP=1939:0100       SS:SP=1939:fffe

1939:0100 66BB00020000     mov  ebx,00000200

dosdebug> t
system state: stopped
AX=0000  BX=0200  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffe  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0346
CS:IP=1939:0106       SS:SP=1939:fffe

1939:0106 6766C70311223344 mov  dword [ebx],44332211

dosdebug> 
system state: stopped
AX=0000  BX=0200  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffe  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0346
CS:IP=1939:010e       SS:SP=1939:fffe

1939:010e FF37             push word [bx]

dosdebug> 
system state: stopped
AX=0000  BX=0200  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffc  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0346
CS:IP=1939:0110       SS:SP=1939:fffc

1939:0110 67FF33           push dword [ebx]

dosdebug> 
system state: stopped
AX=0000  BX=0200  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffa  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0346
CS:IP=1939:0113       SS:SP=1939:fffa

1939:0113 66FF37           push dword [bx]

dosdebug> 
system state: stopped
AX=0000  BX=0200  CX=0000  DX=0000  SI=0000  DI=0000  SP=fff6  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0346
CS:IP=1939:0116       SS:SP=1939:fff6

1939:0116 6766FF33         push dword [ebx]

dosdebug> 
system state: stopped
AX=0000  BX=0200  CX=0000  DX=0000  SI=0000  DI=0000  SP=fff2  BP=0000
DS=1939  ES=1939  FS=0000  GS=0000  FL=000a0346
CS:IP=1939:011a       SS:SP=1939:fff2

1939:011a 90               nop

dosdebug> kill
dosdebug> 
Dosemu process ended - quitting
.dosemu/run$ 
```

**Attach program/game binaries or provide an URL**

[20200520-ldebugu.zip](https://github.com/dosemu2/dosemu2/files/4656319/20200520-ldebugu.zip)

**Attach the log**

[boot.log](https://github.com/dosemu2/dosemu2/files/4656324/boot.log)

**Additional info**

Running the following shell and kernel:

```
C:\>ver /r
 
FreeCom version 0.84-pre7 - GNUC - XMS_Swap [Dec 29 2019 15:36:33]
DOS version 7.10
FreeDOS kernel - SVN (build 2042 OEM:0xfd) [compiled Sep 22 2017]
C:\>
```

**Comment by stsp** [comment]
Maybe something like this will work?
```
diff --git a/src/plugin/debugger/dis8086.c b/src/plugin/debugger/dis8086.c
index 3cf61548c..f0b6fe0dc 100644
--- a/src/plugin/debugger/dis8086.c
+++ b/src/plugin/debugger/dis8086.c
@@ -747,7 +747,7 @@ static void do_modrm(char subtype)
   }
   if (must_do_size) {
     if (wordop) {
-      if (addrsize==32 || opsize==32) {       /* then must specify size */
+      if (addrsize==32) {       /* then must specify size */
                ua_str("dword ");
       } else {
                ua_str("word ");
```
Could you please take a look yourself,
should not be difficult.

**Comment by ecm-pushbx** [comment]
I think your patch is the reverse of what should be done. `opsize==32` indicates that the operand size is `dword`. I will test patching this today or tomorrow.

**Comment by stsp** [comment]
Would be cool, thanks.

**Comment by ecm-pushbx** [comment]
This patch seems to correct it:

```
diff --git a/src/plugin/debugger/dis8086.c b/src/plugin/debugger/dis8086.c
index 3cf61548c..a087b5d62 100644
--- a/src/plugin/debugger/dis8086.c
+++ b/src/plugin/debugger/dis8086.c
@@ -747,7 +747,7 @@ static void do_modrm(char subtype)
   }
   if (must_do_size) {
     if (wordop) {
-      if (addrsize==32 || opsize==32) {       /* then must specify size */
+      if (opsize==32) {       /* then must specify size */
                ua_str("dword ");
       } else {
                ua_str("word ");
```

**Comment by stsp** [comment]
Then please make a PR so that people can review it.

**Comment by ecm-pushbx** [comment]
Added a PR https://github.com/dosemu2/dosemu2/pull/1170

**Comment by stsp** [comment]
Thanks.
Next time add [fixes #1165] for an auto-close.
